axum-extra/src/protobuf.rs

```diff
@@ -0,0 +1,303 @@
+//! Protocol Buffer extractor and response.
+
+use axum::{
+    async_trait,
+    body::{Bytes, HttpBody},
+    extract::{rejection::BytesRejection, FromRequest, RequestParts},
+    response::{IntoResponse, Response},
+    BoxError,
+};
+use http::StatusCode;
+use prost::Message;
+use std::ops::{Deref, DerefMut};
+
+/// Protocol Buffer extractor and response.
+///
+/// When used as an extractor, it can decode request bodies into some type that
+/// implements [`prost::Message`]. The request will be rejected (and a [`ProtoBufRejection`] will
+/// be returned) if:
+///
+/// - The body couldn't be decoded into the target Protocol Buffer.
+/// - Buffering the request body fails.
+///
+/// See [`ProtoBufRejection`] for more details.
+///
+/// # Extractor example
+///
+/// ```rust,no_run
+/// use axum::{routing::post, Router};
+/// use axum_extra::protobuf::ProtoBuf;
+///
+/// #[derive(prost::Message)]
+/// struct CreateUser {
+///     #[prost(string, tag="1")]
+///     email: prost::alloc::string::String,
+///     #[prost(string, tag="2")]
+///     password: prost::alloc::string::String,
+/// }
+///
+/// async fn create_user(ProtoBuf(payload): ProtoBuf<CreateUser>) {
+///     // payload is `CreateUser`
+/// }
+///
+/// let app = Router::new().route("/users", post(create_user));
+/// # async {
+/// # axum::Server::bind(&"".parse().unwrap()).serve(app.into_make_service()).await.unwrap();
+/// # };
+/// ```
+///
+/// When used as a response, it can encode any type that implements [`prost::Message`] to
+/// a newly allocated buffer, and will automatically set `Content-Type: application/octet-stream` header.
```

Comment:
Like the extractor: I think we should put this paragraph into the response section below.